src/Kestrel.Core/KestrelServerOptions.cs

```diff
@@ -55,6 +57,31 @@ public class KestrelServerOptions
         /// </summary>
         public KestrelServerLimits Limits { get; } = new KestrelServerLimits();
 
+        /// <summary>
+        /// Provides a configuration source where endpoints will be loaded from on server start.
+        /// The default is null.
+        /// </summary>
+        public IKestrelConfigBuilder ConfigurationBuilder { get; set; }
+
+        /// <summary>
+        /// A default configuration action for all endpoints. Use for Listen, configuration, the default url, and URLs.
+        /// </summary>
+        internal Action<ListenOptions> EndpointDefaults { get; set; } = _ => { };
+
+        /// <summary>
+        /// Used to flow settings for connection adapters and other extensions.
+        /// </summary>
+        public IDictionary<string, object> AdapterData { get; } = new Dictionary<string, object>(0);
```

Comment:
Yuck. IFeatureCollection?

Reply:
A dictionary in sheep's clothing... You prefer the typed keys?

Reply:
This is pretty gross. How is it used?

Reply:
@davidfowl It's there to allow extension methods defined by connection adapters in separate assemblies to store data between calls to extension methods.

In this case it is used by KestrelServerOptions.ConfigureHttpsDefaults() to store away a default HttpsConnectionAdapterOptions for later use by ListenOptions.UseHttps() and KestrelConfigBuilder.Build().

Reply:
This looks pretty weird, I'm still catching up with this PR and I'll see if I have any better alternatives for what you guys are trying to do